docs/design_spec.md

```diff
@@ -0,0 +1,94 @@
+ is capable of directly configuring that data plane.
+ 
+## Agent Northbound Design
+ 
+Nimbess Agent is able to support multiple container platforms and by default supports Kubernetes (K8S). It
+also supports the CNI spec for handling CNI requests from Kubelet or other container platforms which use
+CNI. The Nimbess Agent NB is responsible for translating K8S resources into network-oriented resources.
+
+The Agent exposes a gRPC based CNI API in order to service CNI requests. The requests are translated into
+Nimbess Ports, and a Network Pipeline is built for that Port. For other networking related container platform
+resources, such as a K8S Service, Ingress, etc; Nimbess Agent will handle translating those resources into their
+relevant network-oriented constructs and applied to one or more pipelines.
+ 
+## Network Constructs in Nimbess Agent
+ 
+In Nimbess, how traffic flows through a data plane can be though of as a pipeline of Network Functions.
+Traffic ingresses a data plane, flows through a series of network functions such as a Switch, ACL,
+Load-Balancer, etc and then egresses the data plane out of a port. Nimbess uses this concept of a Network
+Pipeline and creates one for each Port that is attached to the data plane. There are multiple reasons for
+this design:
+ 
+ 1. Guarantees loop-free forwarding
+ 2. Increases per network function processing efficiency
+ 3. Simplifies debugging
+ 
+When a Port is added to a pipeline, it is broken up into an IngressPort as well as an EgressPort. Only the
+IngressPort is ever attached to this Port's pipeline, while the EgressPort is attached to every other pipeline which
+belongs to the same network fabric (L2/L3 domain). This prevents traffic from ever being forwarded back
+out the same interface from which it originated (1). Additionally, this design increases network
+function processing efficiency, because the cache of the tables are limited to lookups for traffic that may ingress
+a single port. Debugging becomes easier because the only traffic in a particular pipeline only ever comes
+from a single port. However, one of the drawbacks to this design is that there is increased memory usage,
+as the number of instances of network functions scale by the number of ports in the system.
+
+Although each Port contains its own Port Pipeline, the core network functions portion of the pipeline
+remain the same for Ports that belong to the same network fabric. Therefore Nimbess abstracts this portion of
+the Port Pipelines into what is called a Meta Pipeline. This allows Nimbess to make core updates to only
+a single Meta Pipeline, which will then be used to create or update Port Pipelines. Here is an example of
+what the pipelines may look like for a single L2 network with 2 Ports:
+
+[![Nimbess Pipelines](./img/Nimbess_Pipelines.png)](./img/Nimbess_Pipelines.png)
+
+In the example above, the EgressPort for Port 1 is attached to every other Port Pipeline (in this case only 1)
```

Comment:
related to above, this text details that the egress port is dded to each pipeline - does that make it a new pipeline or is the port just something attached to the pipeline since the pipeline is composed of functions. Or is that what you refer to a the metapipeline?

Reply:
egress port is added to every other existing port pipeline. In other words, the other existing port pipelines (belonging to the same network) are updated with the new egress port module attached.